repair/repair.cc

```diff
@@ -995,112 +823,14 @@ static future<> repair_ranges(lw_shared_ptr<repair_info> ri) {
-            ranges = get_primary_ranges_within_dc(db.local(), keyspace);
-        } else if (options.data_centers.size() > 0 || options.hosts.size() > 0) {
-            throw std::runtime_error("You need to run primary range repair on all nodes in the cluster.");
-        } else {
-            ranges = get_primary_ranges(db.local(), keyspace);
-        }
-    } else {
-        ranges = db.local().get_keyspace_local_ranges(keyspace);
-    }
-
-    if (!options.data_centers.empty() && !options.hosts.empty()) {
-        throw std::runtime_error("Cannot combine data centers and hosts options.");
-    }
-
-    if (!options.ignore_nodes.empty() && !options.hosts.empty()) {
-        throw std::runtime_error("Cannot combine ignore_nodes and hosts options.");
-    }
-    std::unordered_set<gms::inet_address> ignore_nodes;
-    for (const auto& n: options.ignore_nodes) {
-        try {
-            auto node = gms::inet_address(n);
-            ignore_nodes.insert(node);
-        } catch(...) {
-            throw std::runtime_error(format("Failed to parse node={} in ignore_nodes={} specified by user: {}",
-                n, options.ignore_nodes, std::current_exception()));
-        }
-    }
-
-    if (!options.start_token.empty() || !options.end_token.empty()) {
-        // Intersect the list of local ranges with the given token range,
-        // dropping ranges with no intersection.
-        std::optional<::range<dht::token>::bound> tok_start;
-        std::optional<::range<dht::token>::bound> tok_end;
-        if (!options.start_token.empty()) {
-            tok_start = ::range<dht::token>::bound(
-                dht::token::from_sstring(options.start_token),
-                false);
-        }
-        if (!options.end_token.empty()) {
-            tok_end = ::range<dht::token>::bound(
-                dht::token::from_sstring(options.end_token),
-                true);
-        }
-        auto wrange = wrapping_range<dht::token>(tok_start, tok_end);
-        dht::token_range_vector given_ranges;
-        ::compat::unwrap_into(std::move(wrange), dht::token_comparator(), [&] (dht::token_range&& x) {
-            given_ranges.push_back(std::move(x));
-        });
-        dht::token_range_vector intersections;
-        for (const auto& range : ranges) {
-            for (const auto& given_range : given_ranges) {
-                auto intersection_opt = range.intersection(given_range, dht::token_comparator());
-                if (intersection_opt) {
-                    intersections.push_back(*intersection_opt);
-                }
-            }
-        }
-        ranges = std::move(intersections);
-    }
-
-    std::vector<sstring> cfs =
-        options.column_families.size() ? options.column_families : list_column_families(db.local(), keyspace);
-    if (cfs.empty()) {
-        rlogger.info("repair[{}]: completed successfully: no tables to repair", id.uuid);
-        return id.id;
-    }
```

Comment:
to improve the patch review, can we refactor the code to keep the code in place here
and call it from the new framework, and then maybe as a second step move it there as is?